torch/nn/modules/module.py

```diff
@@ -193,13 +193,40 @@ def _apply(self, fn):
         for module in self.children():
             module._apply(fn)
 
-        for param in self._parameters.values():
+        def compute_should_use_data(tensor, tensor_applied):
+            if torch._has_same_tensorimpl_type(tensor, tensor_applied):
+                # If the new tensor has the same TensorImpl type as the existing tensor,
+                # then we take `torch.__future__.get_overwrite_module_params_on_conversion()`
+                # into account.
+                return not torch.__future__.get_overwrite_module_params_on_conversion()
+            else:
+                # If the new tensor has a different TensorImpl type, we always overwrite
+                # the existing tensor.
+                return False
+
+        for key, param in self._parameters.items():
             if param is not None:
-                # Tensors stored in modules are graph leaves, and we don't
-                # want to create copy nodes, so we have to unpack the data.
-                param.data = fn(param.data)
-                if param._grad is not None:
-                    param._grad.data = fn(param._grad.data)
+                should_use_data = compute_should_use_data(param, fn(torch.empty(1, dtype=param.dtype, layout=param.layout, device=param.device)))
```

Comment:
this is pretty clever, but I don't think you should do it.  We don't actually guarantee that empty works on all Tensor types (e.g. I don't think it works on quantized).

Instead, what about just adding a parameter for this and just passing in the right answer for all the functions that call `_apply`?  And you can just make the default False for now, I guess.